osidb/sync_manager.py

```diff
@@ -0,0 +1,614 @@
+import json
+from datetime import timedelta
+
+from celery.exceptions import Ignore
+from celery.utils.log import get_task_logger
+from django.conf import settings
+from django.db import models, transaction
+from django.utils import timezone
+
+from config.celery import app
+from osidb.core import set_user_acls
+
+logger = get_task_logger(__name__)
+
+
+class SyncManager(models.Model):
+    """
+    Abstract model to handle synchronization of some OSIDB data with external system like Bugzilla
+    or Jira. Its purpose is to handle scheduling Celery tasks, storing meta-data about when those
+    tasks were processed, and re-scheduling tasks when necessary.
+    """
+
+    MAX_CONSECUTIVE_FAILURES = 5
+    MAX_SCHEDULE_DELAY = timedelta(hours=24)
+    MAX_RUN_LENGTH = timedelta(hours=1)
+    FAIL_RESCHEDULE_DELAY = timedelta(minutes=5)
+
+    sync_id = models.CharField(max_length=100)
+    last_scheduled_dt = models.DateTimeField(blank=True, null=True)
+    last_started_dt = models.DateTimeField(blank=True, null=True)
+    last_finished_dt = models.DateTimeField(blank=True, null=True)
+    last_failed_dt = models.DateTimeField(blank=True, null=True)
+    last_failed_reason = models.TextField(blank=True, null=True)  # noqa: DJ01
```

Comment:
Having `null=True` for char based fields in Django is discouraged, however as discussed in person, it actually makes sense in this scenario as None (NULL) vs empty string differentiate between never encountering any `last_failed_reason` and having `last_failed_reason` actually wrongfully set as empty string by some faulty behaviour.